### DISCUSSION TITLE
Boosts and Favourites counts

### DISCUSSION THREAD

**Discussion by ralfr** [mastodon/mastodon#22627] [Q&A]
Can someone clarify how **exactly** boosts and favourites are counted and federated across Mastodon instances?

I'm running a 35k+ users instance and do get questions related to these counts regularly. From my limited experiments, I assume the following:

- When a post receives a b/f from users local or remote to the post, the b/f is counted and stored with the post **on its home instance** and on the **local copy of the post** on the remote instance of the boosting/favouriting user.
- B/f actions are never federated anywhere else than to the home instance of the post and the local copy of the post on the home instance of a remote user.

As a consequence:

- The total number of b/f is the most accurate only on the home server of the post.
- Copies of a post on remote servers only ever show the total count of b/f from users local to that remote server.
- Therefore, in order to get an accurate number of b/f for any given post, a client would have to reach out to the home instance for that post and get the numbers from there.

Is that 100% accurate?

**Comment by derGali** [comment]
I can't 100% confirm this, but to my knowledge this is exactly how it is.
BUT: what prevents any client (e.g. the original web ui) from fetching the b/f count from the original server if you explicitly "open" the post or by any other way of user interaction? That would be so much worth...

**Reply by renchap** [reply]
This could create a lot of load, "only" to get better numbers.

Also the client (at least the built-in web interface) does not do queries to external hosts, due to privacy reasons. This would need to go through your server, which again creates load.

This is by design on a federated network, and is a hard problem to solve.

**Reply by derGali** [reply]
I understand that a client don't want to do that, because of privacy or other reasons - but I think it could nevertheless?
And don't you get all the replies to the original post if you "open" it, anyhow? The server fetches them, why not the b/f count?

And keep in mind that these are not "only" numbers, but for some people a requirement for using a platform. And if you are on a small or single-person server, the numbers would not get just "better" - currently you simply don't get any.

**Reply by Duvio** [reply]
Recently there has been popular post on Lemmy (the federated Reddit alternative) that complained about certain behaviors by Mastodon like this one and #8301/#20533. Numerous users have agreed and it seemed like there has been a consensus that this appears to be a significant stumbling block preventing Mastodon from achieving greater popularity as an alternative Social Media platform. Given the still persisting rise in prominence of the fediverse, it is essential to focus more on addressing these seemingly minor concerns. Although each individual issue may be deemed an "inconvenience" at most, when combined, they form a substantial collection of supposedly insignificant obstacles that users must constantly navigate around. This, in total, certainly goes beyond being a mere "inconvenience." and it is imperative for us to not do the same mistakes others have made. Namely, we should put ourselves in the perspective of the average brain-dead Twitter user instead of assuming everyone is a tech-freak.

To genuinely encourage broader adoption of the fediverse, especially Mastodon, enhancing user-friendliness is crucial. Initially, some believed that the federation aspect would scare off the general user, limiting the fediverse's appeal to only highly-invested computer enthusiasts as those were the only ones capable of understanding it. However, the current reality, with millions of users embracing the fediverse, proves this wrong. It would be terrbily sad to see this brilliant idea crumble merely due to instances' inability to display likes accurately and other minor issues that programmers are too smug to fix.

**Reply by renchap** [reply]
> It would be terrbily sad to see this brilliant idea crumble merely due to instances' inability to display likes accurately and other minor issues that programmers are too smug to fix.

This is not smugness, but lack of time/people. We are very aware of all the things that could be improved for user experience.

There is one full-time developer working on the project, in addition to Eugen (who has many other responsabilities).

This kind of feature is very complex due to the federation, as I explained. There is a lot of work done to improve people's experience, but the todolist is huge and even only keeping up with day-to-day maintenance is hard at the moment for the core team.

**Reply by Duvio** [reply]
I get that and this project is awesome. It's sad to see it progressing so slowly because of these limitations. I would love to provide PRs, but unfortunately front-end development is not my expertise.

**Reply by PurpleAster** [reply]
I'm not technical and writing from the perspective of a user. My account at @mas.to works well all things considered. Love to all! And perhaps my experience below is related - I don't want to derail this important discussion. 

But just in the past few days a change has occurred where, when I hear a notification of a B/F on one of my posts, they seem to have stopped appearing in my notifications feed. I can see them if I go back to the original post and drill down from there. That's not great because obviously I'm not on Mastodon 24/7 and am not in the habit of scrolling my own posts. 

Given the organic way we find accounts of interest to each other, through actions like boosts and favourites, this behavior hinders a core feature of the platform.

Thank you.

**Reply by bitsondatadev** [reply]
I see this as more of a feature rather than a bug. I understand the argument for meeting the people where they are for increasing adoption, but I can't ignore how wonderful it's been not to think about how myself or others are playing the game for likes.

I do agree that when a developer has time to do the hard (and damn near thankless) work, that having a consistent and elegant design would be nice. In my head, you would have a set of server settings that dictate what the fediverse sees regarding stats, and if individuals would also have the option to disable their own stats from being seen.

This would give the world of business and folks who like to use social media as a metric to success the option to have that world while enabling other instances or individuals who don't want to play the social status games.

The inconsistency itself would remove a lot of the pain that people feel around not feeling seen and just get us back to social interactions on the web again.

My mental health has improved drastically since leaving X and moving to Mastodon and I hope it keeps challenging the norms and reigning in a new era of social interaction. Even if the “feature” is just a lack of implementation, I love it. Thanks to everyone on this project! ❤️

**Reply by snullp** [reply]
I would like to note that for Boosts, instances that have users follow the booster will also be notified so the toot's boost count will +1 there.

**Reply by snullp** [reply]
@renchap I would like to suggest starting to consider this logic as a bug, because of the "Explore" page which does not  exist at the time of the post. The page is driven by those numbers and without an accurate number, instances, especially smaller ones, will have skewed "Explore" page.

**Reply by renchap** [reply]
Some changes were recently merged in `main` where the boosts & like counts are fetched from the remote instance when fetching a post.

**Reply by gelbphoenix** [reply]
So could we see the changes in a possible v4.3.2 release? Or at earliest as the v4.4.0 release (when ever that is)?
